pkg/rmb/rmb.go

```diff
@@ -0,0 +1,251 @@
+	systemLocalBus = "msgbus.system.local"
+	replyBus       = "msgbus.system.reply"
+	numWorkers     = 5
+)
+
+type twinKeyID struct{}
+type messageKey struct{}
+
+// Message is an struct used to communicate over the messagebus
+type Message struct {
+	Version    int      `json:"ver"`
+	UID        string   `json:"uid"`
+	Command    string   `json:"cmd"`
+	Expiration int      `json:"exp"`
+	Retry      int      `json:"try"`
+	Data       string   `json:"dat"`
+	TwinSrc    uint32   `json:"src"`
+	TwinDest   []uint32 `json:"dest"`
+	Retqueue   string   `json:"ret"`
+	Schema     string   `json:"shm"`
+	Epoch      int64    `json:"now"`
+	Err        string   `json:"err"`
+}
+
+// MessageBus is a struct that contains everything required to run the message bus
+type MessageBus struct {
+	Context  context.Context
+	pool     *redis.Pool
+	handlers map[string]func(ctx context.Context, payload []byte) (interface{}, error)
+}
+
+// New creates a new message bus
+func New(ctx context.Context, address string) (*MessageBus, error) {
+	pool, err := newRedisPool(address)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to connect to %s", address)
+	}
+
+	return &MessageBus{
+		pool:     pool,
+		Context:  ctx,
+		handlers: make(map[string]func(ctx context.Context, payload []byte) (interface{}, error)),
+	}, nil
+}
+
+// WithHandler adds a topic handler to the messagebus
+func (m *MessageBus) WithHandler(topic string, handler func(ctx context.Context, payload []byte) (interface{}, error)) {
+	m.handlers[topic] = handler
+}
+
+// Run runs listeners to the configured handlers
+// and will trigger the handlers in the case an event comes in
+func (m *MessageBus) Run() error {
+	con := m.pool.Get()
+	defer con.Close()
+
+	topics := make([]string, len(m.handlers))
+	for topic := range m.handlers {
+		topics = append(topics, topic)
+	}
+
+	jobs := make(chan Message, numWorkers)
+	for i := 1; i <= numWorkers; i++ {
+		go m.worker(context.Background(), jobs)
+	}
+
+	for {
+		if m.Context.Err() != nil {
+			return nil
+		}
+
+		data, err := redis.ByteSlices(con.Do("BLPOP", redis.Args{}.AddFlat(topics).Add(0)...))
+		if err != nil {
+			log.Err(err).Msg("failed to read from system local messagebus")
+			return err
+		}
+
+		var message Message
+		err = json.Unmarshal(data[1], &message)
+		if err != nil {
+			log.Err(err).Msg("failed to unmarshal message")
+			continue
+		}
+
+		_, ok := m.handlers[string(data[0])]
+		if !ok {
+			log.Debug().Msg("handler not found")
+			continue
+		}
+
+		jobs <- message
+	}
+}
+
+func (m *MessageBus) worker(ctx context.Context, jobs chan Message) {
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case message := <-jobs:
+			bytes, err := message.GetPayload()
+			if err != nil {
+				log.Err(err).Msg("err while parsing payload reply")
+			}
+
+			handler, ok := m.handlers[message.Command]
+			if !ok {
+				log.Warn().Msg("handler not found")
+			}
+
+			var tKey twinKeyID
+			ctx = context.WithValue(ctx, tKey, message.TwinSrc)
```

Comment:
this should be 
`requestCtx := context.WithValue(ctx, twinKeyID{}, message.TwinSrc)`
the idea is to create a new context per request. now this is instead overriding the context passed to the function because this is in a loop. so u actually override the ctx each time

Hence the correct way to do it is to create a new context for each request which leaves the ctx passed to the function as it is. this main ctx will be used for cancellation later